src/transformStateWithClones.js

```diff
@@ -6,8 +6,39 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(currentState, action.extraData);
+        stateHistory.push({ ...currentState });
+        break;
+
+      case 'removeProperties':
+        for (const ch of action.keysToRemove) {
+          if (currentState.hasOwnProperty(ch)) {
+            delete currentState[ch];
+          }
```

Comment:
The check for `currentState.hasOwnProperty(ch)` is unnecessary because the `delete` operator will not throw an error if the property does not exist. You can safely remove this check.